File: packages/journey/ios/Callback/JourneyCallbackValueApplier.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingJourney
import PingJourneyPlugin
import PingOrchestrate

/// Applies callback input mutations from JS payloads to active Journey callbacks.
enum JourneyCallbackValueApplier {
  /// Callback types that require additional integration before values can be mutated.
  private static let integrationCallbackRequirements: [String: String] = [
    "DeviceProfileCallback": "@react-native-pingidentity/device-profile",
    "PingOneProtectInitializeCallback": "PingOne Protect integration",
    "PingOneProtectEvaluationCallback": "PingOne Protect integration",
    "Fido2RegistrationCallback": "FIDO/WebAuthn integration",
    "Fido2AuthenticationCallback": "FIDO/WebAuthn integration",
    "FidoRegistrationCallback": "FIDO/WebAuthn integration",
    "FidoAuthenticationCallback": "FIDO/WebAuthn integration",
    "SelectIdPCallback": "External IdP integration",
    "IdPCallback": "External IdP integration",
    "RedirectCallback": "Redirect handling integration",
    "ReCaptchaCallback": "ReCaptcha integration",
    "ReCaptchaEnterpriseCallback": "ReCaptcha Enterprise integration",
    "BindingCallback": "Binding integration",
    "DeviceBindingCallback": "Binding integration",
    "DeviceSigningVerifierCallback": "Binding integration"
  ]

  /// Callback types that are output-only and cannot be mutated via `next()`.
  private static let outputOnlyCallbackTypes: Set<String> = [
    "MetadataCallback",
    "PollingWaitCallback",
    "TextOutputCallback",
    "SuspendedTextOutputCallback"
  ]

  /// Parsed callback mutation instruction received from JavaScript.
  ///
  /// - Note: `@unchecked Sendable` is used because `value` is `Any?`, which
  ///   is not statically sendable. Parsed mutations are immutable value types
  ///   and consumed synchronously within native mutation flow.
  struct CallbackMutation: @unchecked Sendable {
    let type: String
    let value: Any?
    let index: Int?
  }

  /// Parses callback mutation payload from bridge input.
  ///
  /// - Parameter input: Bridge input map.
  /// - Returns: Parsed callback mutations.
  /// - Throws: `JourneyBridgeError.argument` when payload is malformed.
  static func parseInput(_ input: NSDictionary) throws -> [CallbackMutation] {
    guard let callbacks = input["callbacks"] as? [Any] else {
      return []
    }

    var mutations = [CallbackMutation]()
    for (index, item) in callbacks.enumerated() {
      guard let callbackMap = item as? NSDictionary else {
        throw JourneyBridgeError.argument("Invalid callback payload at index \(index)")
      }
      guard let type = callbackMap["type"] as? String, !type.isEmpty else {
        throw JourneyBridgeError.argument("Callback type is required at index \(index)")
      }

      mutations.append(
        CallbackMutation(
          type: type,
          value: readDynamicValue(callbackMap["value"]),
          index: parseOptionalIndex(callbackMap["index"])
        )
      )
    }

    return mutations
  }

  /// Applies parsed callback mutations to an active continue node.
  ///
  /// - Parameters:
  ///   - continueNode: Active continue node.
  ///   - mutations: Parsed callback mutations.
  /// - Throws: `JourneyBridgeError` when mutation fails.
  static func apply(_ continueNode: ContinueNode, mutations: [CallbackMutation]) throws {
    try applyToCallbacks(continueNode.callbacks.map { $0 }, mutations: mutations)
  }

  /// Applies parsed callback mutations to callback list.
  ///
  /// - Parameters:
  ///   - callbacks: Active callbacks.
  ///   - mutations: Parsed callback mutations.
  /// - Throws: `JourneyBridgeError` when mutation fails.
  static func applyToCallbacks(_ callbacks: [Any], mutations: [CallbackMutation]) throws {
    var consumedIndexByType = [String: Int]()

    for mutation in mutations {
      let callback = try findCallback(
        callbacks: callbacks,
        mutation: mutation,
        consumedIndexByType: &consumedIndexByType
      )
      let value = mutation.value

      switch callback {
      case let callback as NameCallback:
        callback.name = try asString(value, fieldName: mutation.type)

      case let callback as PasswordCallback:
        callback.password = try asString(value, fieldName: mutation.type)

      case let callback as TextInputCallback:
        callback.text = try asString(value, fieldName: mutation.type)

      case let callback as StringAttributeInputCallback:
        callback.value = try asString(value, fieldName: mutation.type)

      case let callback as NumberAttributeInputCallback:
        callback.value = try asDouble(value, fieldName: mutation.type)

      case let callback as BooleanAttributeInputCallback:
        callback.value = try asBoolean(value, fieldName: mutation.type)

      case let callback as HiddenValueCallback:
        callback.value = try asString(value, fieldName: mutation.type)

      case let callback as TermsAndConditionsCallback:
        callback.accepted = try asBoolean(value, fieldName: mutation.type)

      case let callback as ConsentMappingCallback:
        callback.accepted = try asBoolean(value, fieldName: mutation.type)

      case let callback as ChoiceCallback:
        callback.selectedIndex = try asInt(value, fieldName: mutation.type)

      case let callback as ConfirmationCallback:
        callback.selectedIndex = try asInt(value, fieldName: mutation.type)

      case let callback as ValidatedPasswordCallback:
        callback.password = try asString(value, fieldName: mutation.type)

      case let callback as ValidatedUsernameCallback:
        callback.username = try asString(value, fieldName: mutation.type)

      case let callback as KbaCreateCallback:
        let map = asStringMap(value)
        if let selectedQuestion = map["selectedQuestion"] ?? map["question"] {
          callback.selectedQuestion = try asString(
            selectedQuestion,
            fieldName: "\(mutation.type).selectedQuestion"
          )
        }
        if let selectedAnswer = map["selectedAnswer"] ?? map["answer"] {
          callback.selectedAnswer = try asString(
            selectedAnswer,
            fieldName: "\(mutation.type).selectedAnswer"
          )
        } else if let stringValue = value as? String {
          callback.selectedAnswer = stringValue
        }
        if let allowUserDefinedQuestions = map["allowUserDefinedQuestions"] {
          callback.allowUserDefinedQuestions = try asBoolean(
            allowUserDefinedQuestions,
            fieldName: "\(mutation.type).allowUserDefinedQuestions"
          )
        }

      case is MetadataCallback, is PollingWaitCallback, is TextOutputCallback, is SuspendedTextOutputCallback:
        throw JourneyBridgeError.unsupportedCallback(
          "Callback type \(mutation.type) is output-only and cannot be mutated"
        )

      default:
        let normalizedType = normalizedCallbackType(mutation.type)
        if outputOnlyCallbackTypes.contains(normalizedType) {
          throw JourneyBridgeError.unsupportedCallback(
            "Callback type \(mutation.type) is output-only and cannot be mutated"
          )
        }
        if let requirement = integrationCallbackRequirements[normalizedType] {
          throw JourneyBridgeError.missingIntegration(
            "Callback type \(mutation.type) requires additional native integration: \(requirement)"
          )
        }
        throw JourneyBridgeError.unsupportedCallback(
          "Callback type \(mutation.type) is not supported for value mutation"
        )
      }
    }
  }

  /// Resolves one callback target for a mutation, accounting for callback aliases and indexes.
  ///
  /// - Parameters:
  ///   - callbacks: Active callbacks.
  ///   - mutation: Parsed callback mutation.
  ///   - consumedIndexByType: Mutable sequential index tracker keyed by callback type.
  /// - Returns: Matching callback instance.
  /// - Throws: `JourneyBridgeError.argument` when no matching callback is found.
  private static func findCallback(
    callbacks: [Any],
    mutation: CallbackMutation,
    consumedIndexByType: inout [String: Int]
  ) throws -> Any {
    let normalizedType = normalizedCallbackType(mutation.type)
    let matching = callbacks.filter { callback in
      callbackLookupType(callback) == normalizedType
    }

    if matching.isEmpty {
      throw JourneyBridgeError.argument("No active callback found for type \(mutation.type)")
    }

    let index = mutation.index ?? consumedIndexByType[normalizedType, default: 0]
    consumedIndexByType[normalizedType] = index + 1

    guard index >= 0, index < matching.count else {
      throw JourneyBridgeError.argument(
        "No active callback found for type \(mutation.type) at index \(index)"
      )
    }
    return matching[index]
  }

  /// Resolves callback lookup type from runtime callback and callback JSON payload.
  ///
  /// - Parameter callback: Runtime callback instance.
  /// - Returns: Normalized callback type.
  private static func callbackLookupType(_ callback: Any) -> String {
    if let abstract = callback as? AbstractCallback,
       let rawType = abstract.json["type"] as? String,
       !rawType.isEmpty {
      return normalizedCallbackType(rawType)
    }
    return normalizedCallbackType(String(describing: type(of: callback)))
  }

  /// Normalizes alias callback type names used by JS helper APIs.
  ///
  /// - Parameter type: Callback type from mutation payload.
  /// - Returns: Native callback type name used in callback lookup.
  private static func normalizedCallbackType(_ type: String) -> String {
    switch type {
    case "ValidatedCreatePasswordCallback":
      return "ValidatedPasswordCallback"
    case "ValidatedCreateUsernameCallback":
      return "ValidatedUsernameCallback"
    default:
      return type
    }
  }

  /// Parses optional callback index from bridge payload.
  ///
  /// - Parameter rawValue: Raw index field.
  /// - Returns: Parsed callback index or `nil`.
  private static func parseOptionalIndex(_ rawValue: Any?) -> Int? {
    if let number = rawValue as? NSNumber {
      return number.intValue
    }
    if let string = rawValue as? String {
      return Int(string)
    }
    return nil
  }

  /// Converts bridge values to optional dynamic values.
  ///
  /// - Parameter rawValue: Raw bridge value.
  /// - Returns: Dynamic value.
  private static func readDynamicValue(_ rawValue: Any?) -> Any? {
    if rawValue is NSNull {
      return nil
    }
    return rawValue
  }

  /// Coerces a dynamic value to a string.
  ///
  /// - Parameters:
  ///   - value: Dynamic value.
  ///   - fieldName: Field path used in validation errors.
  /// - Returns: String value.
  /// - Throws: `JourneyBridgeError.argument` when value cannot be represented as a string.
  private static func asString(_ value: Any?, fieldName: String) throws -> String {
    switch value {
    case let string as String:
      return string
    case let number as NSNumber:
      return number.stringValue
    case let bool as Bool:
      return bool ? "true" : "false"
    default:
      throw JourneyBridgeError.argument("\(fieldName) expects a string-compatible value")
    }
  }

  /// Coerces a dynamic value to a boolean.
  ///
  /// - Parameters:
  ///   - value: Dynamic value.
  ///   - fieldName: Field path used in validation errors.
  /// - Returns: Boolean value.
  /// - Throws: `JourneyBridgeError.argument` when value cannot be represented as a boolean.
  private static func asBoolean(_ value: Any?, fieldName: String) throws -> Bool {
    switch value {
    case let bool as Bool:
      return bool
    case let number as NSNumber:
      return number.intValue != 0
    case let string as String:
      let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
      if normalized == "true" || normalized == "1" {
        return true
      }
      if normalized == "false" || normalized == "0" {
        return false
      }
      throw JourneyBridgeError.argument("\(fieldName) expects a boolean value")
    default:
      throw JourneyBridgeError.argument("\(fieldName) expects a boolean value")
    }
  }

  /// Coerces a dynamic value to a double.
  ///
  /// - Parameters:
  ///   - value: Dynamic value.
  ///   - fieldName: Field path used in validation errors.
  /// - Returns: Double value.
  /// - Throws: `JourneyBridgeError.argument` when value is not numeric.
  private static func asDouble(_ value: Any?, fieldName: String) throws -> Double {
    switch value {
    case let number as NSNumber:
      return number.doubleValue
    case let string as String:
      let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines)
      if let parsed = Double(normalized) {
        return parsed
      }
      throw JourneyBridgeError.argument("\(fieldName) expects a numeric value")
    default:
      throw JourneyBridgeError.argument("\(fieldName) expects a numeric value")
    }
  }

  /// Coerces a dynamic value to an integer.
  ///
  /// - Parameters:
  ///   - value: Dynamic value.
  ///   - fieldName: Field path used in validation errors.
  /// - Returns: Integer value.
  /// - Throws: `JourneyBridgeError.argument` when value is not an integer.
  private static func asInt(_ value: Any?, fieldName: String) throws -> Int {
    switch value {
    case let number as NSNumber:
      return number.intValue
    case let string as String:
      let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines)
      if let parsed = Int(normalized) {
        return parsed
      }
      throw JourneyBridgeError.argument("\(fieldName) expects an integer value")
    default:
      throw JourneyBridgeError.argument("\(fieldName) expects an integer value")
    }
  }

  /// Coerces a dynamic value to a string-keyed map.
  ///
  /// - Parameter value: Dynamic value.
  /// - Returns: String-keyed map representation.
  private static func asStringMap(_ value: Any?) -> [String: Any] {
    guard let dictionary = value as? NSDictionary else {
      return [:]
    }
    var mapped = [String: Any]()
    dictionary.forEach { key, item in
      guard let key = key as? String else {
        return
      }
      mapped[key] = item
    }
    return mapped
  }
}

File: packages/journey/android/src/main/java/com/pingidentity/rnjourney/callback/JourneyCallbackValueApplier.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rnjourney

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.pingidentity.journey.callback.BooleanAttributeInputCallback
import com.pingidentity.journey.callback.ChoiceCallback
import com.pingidentity.journey.callback.ConfirmationCallback
import com.pingidentity.journey.callback.ConsentMappingCallback
import com.pingidentity.journey.callback.HiddenValueCallback
import com.pingidentity.journey.callback.KbaCreateCallback
import com.pingidentity.journey.callback.MetadataCallback
import com.pingidentity.journey.callback.NameCallback
import com.pingidentity.journey.callback.NumberAttributeInputCallback
import com.pingidentity.journey.callback.PasswordCallback
import com.pingidentity.journey.callback.PollingWaitCallback
import com.pingidentity.journey.callback.StringAttributeInputCallback
import com.pingidentity.journey.callback.SuspendedTextOutputCallback
import com.pingidentity.journey.callback.TermsAndConditionsCallback
import com.pingidentity.journey.callback.TextInputCallback
import com.pingidentity.journey.callback.TextOutputCallback
import com.pingidentity.journey.callback.ValidatedPasswordCallback
import com.pingidentity.journey.callback.ValidatedUsernameCallback
import com.pingidentity.journey.plugin.callbacks
import com.pingidentity.orchestrate.ContinueNode

/**
 * Applies callback input mutations from JS payloads to active Journey callbacks.
 */
internal object JourneyCallbackValueApplier {
    internal class MissingIntegrationException(message: String) : IllegalStateException(message)

    /**
     * Callback types that require external module integration before values can be applied.
     *
     * Map value describes the required integration for error messaging.
     */
    private val integrationCallbackRequirements = mapOf(
        "DeviceProfileCallback" to "@ping-identity/rn-device-profile",
        "FidoRegistrationCallback" to "@ping-identity/rn-fido",
        "FidoAuthenticationCallback" to "@ping-identity/rn-fido",
        "PingOneProtectInitializeCallback" to "PingOne Protect integration",
        "PingOneProtectEvaluationCallback" to "PingOne Protect integration",
        "SelectIdPCallback" to "External IdP integration",
        "IdPCallback" to "External IdP integration",
        "ReCaptchaCallback" to "ReCaptcha integration",
        "ReCaptchaEnterpriseCallback" to "ReCaptcha Enterprise integration",
        "BindingCallback" to "Binding integration",
        "DeviceBindingCallback" to "Binding integration",
        "DeviceSigningVerifierCallback" to "Binding integration"
    )

    /**
     * Parsed callback mutation instruction received from JavaScript.
     *
     * @property type Callback type name to target.
     * @property value Dynamic callback value to apply.
     * @property index Optional per-type callback index.
     */
    internal data class CallbackMutation(
        val type: String,
        val value: Any?,
        val index: Int?
    )

    /**
     * Parse callback mutation payload from bridge input.
     *
     * @param input Bridge input map.
     * @return Parsed callback mutations.
     */
    fun parseInput(input: ReadableMap): List<CallbackMutation> {
        val callbacks = input.getArray("callbacks") ?: return emptyList()
        val mutations = mutableListOf<CallbackMutation>()

        for (i in 0 until callbacks.size()) {
            val callbackMap = callbacks.getMap(i)
                ?: throw IllegalArgumentException("Invalid callback payload at index $i")
            val type = callbackMap.getString("type")
                ?: throw IllegalArgumentException("Callback type is required at index $i")
            val index = parseOptionalIndex(callbackMap)
            mutations.add(
                CallbackMutation(
                    type = type,
                    value = readDynamicValue(callbackMap, "value"),
                    index = index
                )
            )
        }

        return mutations
    }

    /**
     * Apply parsed callback mutations to the current continue node callbacks.
     *
     * @param continueNode Active continue node.
     * @param mutations Parsed callback mutations.
     */
    suspend fun apply(continueNode: ContinueNode, mutations: List<CallbackMutation>) {
        val callbacks = continueNode.callbacks.map { it as Any }
        applyToCallbacks(callbacks, mutations)
    }

    /**
     * Apply parsed callback mutations to callback list.
     *
     * @param callbacks Active callbacks.
     * @param mutations Parsed callback mutations.
     */
    suspend fun applyToCallbacks(callbacks: List<Any>, mutations: List<CallbackMutation>) {
        val callbacksByType = callbacks.groupBy { it::class.java.simpleName }
        val consumedIndexByType = mutableMapOf<String, Int>()

        mutations.forEach { mutation ->
            val callback = findCallback(callbacksByType, mutation, consumedIndexByType)
            val value = mutation.value

            when (callback) {
                is NameCallback -> callback.name = asString(value, mutation.type)
                is PasswordCallback -> callback.password = asString(value, mutation.type)
                is TextInputCallback -> callback.text = asString(value, mutation.type)
                is StringAttributeInputCallback -> callback.value = asString(value, mutation.type)
                is NumberAttributeInputCallback -> callback.value = asDouble(value, mutation.type)
                is BooleanAttributeInputCallback -> callback.value = asBoolean(value, mutation.type)
                is HiddenValueCallback -> callback.value = asString(value, mutation.type)
                is TermsAndConditionsCallback -> callback.accepted = asBoolean(value, mutation.type)
                is ConsentMappingCallback -> callback.accepted = asBoolean(value, mutation.type)
                is ChoiceCallback -> callback.selectedIndex = asInt(value, mutation.type)
                is ConfirmationCallback -> callback.selectedIndex = asInt(value, mutation.type)
                is ValidatedPasswordCallback -> callback.password = asString(value, mutation.type)
                is ValidatedUsernameCallback -> callback.username = asString(value, mutation.type)
                is KbaCreateCallback -> {
                    val map = asStringMap(value)
                    val selectedQuestion = map["selectedQuestion"] ?: map["question"]
                    val selectedAnswer = map["selectedAnswer"] ?: map["answer"]
                    val allowUserDefinedQuestions = map["allowUserDefinedQuestions"]

                    if (selectedQuestion != null) {
                        callback.selectedQuestion = asString(selectedQuestion, "${mutation.type}.selectedQuestion")
                    }
                    if (selectedAnswer != null) {
                        callback.selectedAnswer = asString(selectedAnswer, "${mutation.type}.selectedAnswer")
                    } else if (value is String) {
                        callback.selectedAnswer = value
                    }
                    if (allowUserDefinedQuestions != null) {
                        callback.allowUserDefinedQuestions = asBoolean(
                            allowUserDefinedQuestions,
                            "${mutation.type}.allowUserDefinedQuestions"
                        )
                    }
                }
                is MetadataCallback,
                is PollingWaitCallback,
                is TextOutputCallback,
                is SuspendedTextOutputCallback -> {
                    throw UnsupportedOperationException(
                        "Callback type ${mutation.type} is output-only and cannot be mutated"
                    )
                }
                else -> {
                    val normalizedType = normalizedCallbackType(mutation.type)
                    val requirement = integrationCallbackRequirements[normalizedType]
                    if (requirement != null) {
                        throw MissingIntegrationException(
                            "Callback type ${mutation.type} requires additional native integration: $requirement"
                        )
                    }

                    throw UnsupportedOperationException(
                        "Callback type ${mutation.type} is not supported for value mutation"
                    )
                }
            }
        }
    }

    /**
     * Normalizes alias callback type names used by JS helper APIs.
     *
     * @param type Callback type from mutation payload.
     * @return Native callback type name used in runtime callback lookup.
     */
    private fun normalizedCallbackType(type: String): String {
        return when (type) {
            "ValidatedCreatePasswordCallback" -> "ValidatedPasswordCallback"
            "ValidatedCreateUsernameCallback" -> "ValidatedUsernameCallback"
            else -> type
        }
    }

    /**
     * Resolves one target callback for a mutation, accounting for callback aliases and indexes.
     *
     * @param callbacksByType Active callbacks grouped by runtime type name.
     * @param mutation Parsed callback mutation.
     * @param consumedIndexByType Running index map used for sequential mutations.
     * @return Matching callback instance.
     * @throws IllegalArgumentException when no matching callback is found.
     */
    private fun findCallback(
        callbacksByType: Map<String, List<Any>>,
        mutation: CallbackMutation,
        consumedIndexByType: MutableMap<String, Int>
    ): Any {
        val normalizedType = normalizedCallbackType(mutation.type)
        val matching = callbacksByType[normalizedType].orEmpty()
        if (matching.isEmpty()) {
            throw IllegalArgumentException("No active callback found for type ${mutation.type}")
        }

        val index = mutation.index ?: consumedIndexByType.getOrDefault(normalizedType, 0)
        consumedIndexByType[normalizedType] = index + 1

        return matching.getOrNull(index)
            ?: throw IllegalArgumentException(
                "No active callback found for type ${mutation.type} at index $index"
            )
    }

    /**
     * Parses optional callback index from bridge payload.
     *
     * @param callbackMap One callback mutation map from JavaScript.
     * @return Parsed callback index or null when absent/invalid.
     */
    private fun parseOptionalIndex(callbackMap: ReadableMap): Int? {
        if (!callbackMap.hasKey("index") || callbackMap.isNull("index")) {
            return null
        }
        return when (callbackMap.getType("index")) {
            ReadableType.Number -> callbackMap.getDouble("index").toInt()
            ReadableType.String -> callbackMap.getString("index")?.toIntOrNull()
            else -> null
        }
    }

    /**
     * Reads one dynamic value from a bridge map using runtime type inspection.
     *
     * @param map Source readable map.
     * @param key Field key.
     * @return Parsed dynamic value.
     */
    private fun readDynamicValue(map: ReadableMap, key: String): Any? {
        if (!map.hasKey(key) || map.isNull(key)) {
            return null
        }
        return when (map.getType(key)) {
            ReadableType.Null -> null
            ReadableType.Boolean -> map.getBoolean(key)
            ReadableType.Number -> map.getDouble(key)
            ReadableType.String -> map.getString(key)
            ReadableType.Map -> map.getMap(key)?.toHashMap()
            ReadableType.Array -> readArrayValue(map.getArray(key))
        }
    }

    /**
     * Converts bridge arrays to plain Kotlin lists.
     *
     * @param array Source readable array.
     * @return Parsed list value.
     */
    private fun readArrayValue(array: ReadableArray?): List<Any?> {
        if (array == null) {
            return emptyList()
        }
        val values = mutableListOf<Any?>()
        for (i in 0 until array.size()) {
            val value = when (array.getType(i)) {
                ReadableType.Null -> null
                ReadableType.Boolean -> array.getBoolean(i)
                ReadableType.Number -> array.getDouble(i)
                ReadableType.String -> array.getString(i)
                ReadableType.Map -> array.getMap(i)?.toHashMap()
                ReadableType.Array -> readArrayValue(array.getArray(i))
            }
            values.add(value)
        }
        return values
    }

    /**
     * Coerces a dynamic value to string.
     *
     * @param value Dynamic value to coerce.
     * @param fieldName Field path used for validation errors.
     * @return String value.
     * @throws IllegalArgumentException when value cannot be represented as a string.
     */
    private fun asString(value: Any?, fieldName: String): String {
        return when (value) {
            is String -> value
            is Number -> value.toString()
            is Boolean -> value.toString()
            else -> throw IllegalArgumentException("$fieldName expects a string-compatible value")
        }
    }

    /**
     * Coerces a dynamic value to boolean.
     *
     * @param value Dynamic value to coerce.
     * @param fieldName Field path used for validation errors.
     * @return Boolean value.
     * @throws IllegalArgumentException when value cannot be represented as a boolean.
     */
    private fun asBoolean(value: Any?, fieldName: String): Boolean {
        return when (value) {
            is Boolean -> value
            is Number -> value.toInt() != 0
            is String -> {
                val normalized = value.trim().lowercase()
                when (normalized) {
                    "true", "1" -> true
                    "false", "0" -> false
                    else -> throw IllegalArgumentException("$fieldName expects a boolean value")
                }
            }
            else -> throw IllegalArgumentException("$fieldName expects a boolean value")
        }
    }

    /**
     * Coerces a dynamic value to double.
     *
     * @param value Dynamic value to coerce.
     * @param fieldName Field path used for validation errors.
     * @return Numeric double value.
     * @throws IllegalArgumentException when value is not numeric.
     */
    private fun asDouble(value: Any?, fieldName: String): Double {
        return when (value) {
            is Number -> value.toDouble()
            is String -> value.toDoubleOrNull()
                ?: throw IllegalArgumentException("$fieldName expects a numeric value")
            else -> throw IllegalArgumentException("$fieldName expects a numeric value")
        }
    }

    /**
     * Coerces a dynamic value to integer.
     *
     * @param value Dynamic value to coerce.
     * @param fieldName Field path used for validation errors.
     * @return Integer value.
     * @throws IllegalArgumentException when value is not an integer.
     */
    private fun asInt(value: Any?, fieldName: String): Int {
        return when (value) {
            is Number -> value.toInt()
            is String -> value.toIntOrNull()
                ?: throw IllegalArgumentException("$fieldName expects an integer value")
            else -> throw IllegalArgumentException("$fieldName expects an integer value")
        }
    }

    /**
     * Coerces a dynamic value to a string-keyed map.
     *
     * @param value Dynamic value to coerce.
     * @return String-keyed map representation, or an empty map.
     */
    private fun asStringMap(value: Any?): Map<String, Any?> {
        val raw = value as? Map<*, *> ?: return emptyMap()
        return raw.entries.mapNotNull { entry ->
            val key = entry.key as? String ?: return@mapNotNull null
            key to entry.value
        }.toMap()
    }
}
